Add missing fields

daiagi opened the pull request:
fork from https://github.com/daiagi/stick/tree/tokenEnityt/fix-count-calc

implement https://github.com/kodadot/loligo/issues/51 for stick

specifically,
1. add chepeastNft field to tokenEntity
2. listing event handler to sync cheapest nft
3. resolver for TokenEntities (with correct count) of specific owner (for "own" filter in UI)
4. add meta and metadata

vikiival commented on db/migrations/1689623692588-Data.js:
Never alter migration pls

daiagi replied:
ok

vikiival commented on schema.graphql:
```diff
@@ -30,14 +30,18 @@ type TokenEntity @entity {
   id: ID!
   blockNumber: BigInt
   collection: CollectionEntity
+  cheapestNFT: NFTEntity
```
```suggestion
  cheapest: NFTEntity
```

vikiival commented on src/mappings/nfts/burn.ts:
```diff
@@ -31,6 +32,8 @@ export async function handleTokenBurn(context: Context): Promise<void> {
   entity.collection.ownerCount = ownerCount
   entity.collection.distribution = distribution
 
+  await eventHandlers.burnHandler(context, entity)
```
-_-

daiagi replied:
-_- ?

vikiival replied:
see point 1.

daiagi replied:
understood

vikiival commented on src/mappings/nfts/list.ts:
```diff
@@ -22,6 +23,7 @@ export async function handleTokenList(context: Context): Promise<void> {
   if (entity.price && (entity.collection.floor === 0n || entity.price < entity.collection.floor)) {
     entity.collection.floor = entity.price
   }
+  eventHandlers.listHandler(context, entity)
```
👀

daiagi replied:
'await' right?

vikiival replied:
right

vikiival commented on src/mappings/shared/handleTokenEntity.ts:
move to `src/mappings/shared/tokens`

vikiival commented on src/mappings/shared/handleTokenEntity.ts:
```diff
@@ -1,44 +1,166 @@
 
-export async function handleTokenEntity(context: Context, collection: CE, nft: NE): Promise<TE | undefined> {
-  const nftMedia = nft.image || nft.media
+function generateTokenId(collectionId: string, nftMedia: string): string {
+  return `${collectionId}-${md5(nftMedia)}`
+}
+
```
move to `src/mappings/shared/tokens/utils.ts`

daiagi replied:
yes sir

vikiival commented on src/mappings/shared/handleTokenEntity.ts:
```diff
@@ -1,44 +1,166 @@
+  const nftMedia = nft.image ?? nft.media
   if (!nftMedia || nftMedia === '') {
     warn(OPERATION, `MISSING NFT MEDIA ${nft.id}`)
     return
```
```suggestion
    return undefined
```

vikiival commented on src/mappings/shared/handleTokenEntity.ts:
```diff
@@ -1,44 +1,166 @@
   }
+  const tokenId = generateTokenId(collection.id, nftMedia)
+  debug(OPERATION, { createToken: `Create TOKEN ${tokenId} for NFT ${nft.id}` })
+  const tokenName = typeof nft.name === 'string' ? nft.name?.replace(/([#_]\d+$)/g, '').trim() : ''
```
move to `src/mappings/shared/tokens/utils.ts`

vikiival commented on src/mappings/shared/handleTokenEntity.ts:
```diff
@@ -1,44 +1,166 @@
-  }
+  nft.token = token
+  await context.store.save(token)
+  await context.store.save(nft)
```
any particular reason to `save` nft here? 
cause slowest op is

daiagi replied:
it needs to be updated with new token
but indeed, update is more suitable and faster

vikiival commented on src/mappings/shared/handleTokenEntity.ts:
```diff
@@ -1,44 +1,166 @@
+}
+
+async function mintHandler(context: Context, collection: CE, nft: NE): Promise<TE | undefined> {
+  const nftMedia = nft.image ?? nft.media
```
move to `src/mappings/shared/tokens/utils.ts`

vikiival commented on src/mappings/shared/handleTokenEntity.ts:
```diff
@@ -1,44 +1,166 @@
+
+  if (!nftMedia || nftMedia === '') {
+    warn(OPERATION, `MISSING NFT MEDIA ${nft.id}`)
+    return
```
```suggestion
    return undefined
```

vikiival commented on src/mappings/shared/handleTokenEntity.ts:
```diff
@@ -1,44 +1,166 @@
+
+async function handleMetadataSet(context: Context, collection: CE, nft: NE): Promise<TE | undefined> {
+  debug(OPERATION, { handleMetadataSet: `Handle set metadata for NFT ${nft.id}` })
+  const nftMedia = nft.image ?? nft.media
+
+  if (!nftMedia || nftMedia === '') {
```
can part of one function in 
`src/mappings/shared/tokens/utils.ts`

vikiival commented on src/mappings/shared/handleTokenEntity.ts:
```diff
@@ -1,44 +1,166 @@
+  await context.store.update(TE, token.id, { count: token.count - 1, updatedAt: nft.updatedAt })
+}
+
+async function listHandler(context: Context, nft: NE): Promise<TE | undefined> {
```
When someone does buy it is not handled

vikiival commented on src/mappings/shared/handleTokenEntity.ts:
```diff
@@ -1,44 +1,166 @@
+}
+
+async function listHandler(context: Context, nft: NE): Promise<TE | undefined> {
+  const nftMedia = nft.image ?? nft.media
+
+  if (!nftMedia || nftMedia === '') {
+    warn(OPERATION, `MISSING NFT MEDIA ${nft.id}`)
+    return
+  }
```
Same as before

vikiival commented on src/mappings/uniques/list.ts:
```diff
@@ -22,6 +23,7 @@ export async function handleTokenList(context: Context): Promise<void> {
   if (entity.price && (entity.collection.floor === 0n || entity.price < entity.collection.floor)) {
     entity.collection.floor = entity.price
   }
+  eventHandlers.listHandler(context, entity)
```
forgot to await, app will crash

vikiival commented on src/server-extension/query/tokenEntityByOwner.ts:
```diff
@@ -0,0 +1,48 @@
+export const tokenEntityByOwner = `WITH CheapestNFT AS (
```
```suggestion
export const tokenEntityByOwner = `WITH cheapest_nft AS (
```

vikiival commented on src/server-extension/query/tokenEntityByOwner.ts:
```diff
@@ -0,0 +1,48 @@
+export const tokenEntityByOwner = `WITH CheapestNFT AS (
+    SELECT 
+        ne.token_id,
+        MIN(ne.price) AS "cheapestNFTPrice"
```
```suggestion
        MIN(ne.price) AS "cheapest"
```

vikiival commented on src/server-extension/query/tokenEntityByOwner.ts:
```diff
@@ -0,0 +1,48 @@
+    t.image as image,
+    t.media as media,
+    t.metadata as metadata,
+    me.id as "metaId",
+    me.description as "metaDescription",
+    me.image as "metaImage",
+    me.animation_url as "metaAnimationUrl",
+    t.block_number as "blockNumber",
```
any camelcase is ignored by resolver

daiagi replied:
it actually isn't when you put it in " "
I wanted to keep the names of fields as they are on TokenEntity

![image](https://github.com/kodadot/stick/assets/22791238/b1ee6af6-388c-4c99-98e8-6d1f1e2f6b35)

vikiival commented on src/server-extension/query/tokenEntityByOwner.ts:
```diff
@@ -0,0 +1,48 @@
+    
+    col.id AS "collectionId",
+    col.name AS "collectionName"
+FROM
+    token_entity as t
+    LEFT JOIN nft_entity as ne ON t.id = ne.token_id AND ne.current_owner = $1
+    LEFT JOIN CheapestNFT as c ON t.id = c.token_id
+    LEFT JOIN collection_entity as col ON t.collection_id = col.id
+    LEFT JOIN metadata_entity as me ON t.meta_id = me.id
```
4 left joins is not a performance wise

daiagi replied:
I have benchmarked a few queries, some with 3 and some with 4 joins

tested by a bash script that runs an SQL query 200 times against the fully loaded DB on localhost

results are, with no significant difference between the approaches

mean between 1.55 to 1.65 ms
standard deviation  ~0.25 ms

vikiival commented on src/server-extension/resolvers/tokenEntityByOwner.ts:
?

daiagi replied:
are you asking about
name?
why it exists?

vikiival reviewed (changes requested):
Hello, thanks for the implementation I see that you tried to do your best!

I have a few main issues:

### 1. Code structure

as we spoke before I try to keep some code structure.
I suggest create `src/mappings/shared/tokens`
and create file per each handler: `create.ts, list.ts ` etc.

2. you forget awaits

`eventHandlers.listHandler(context, entity)` will make the indexer fail in the next block because connection will still locked by this

3. too much repetitive code

I see that some simple functions are often repeated a lot.
When someone decides to refactor the code. he/she can fail so this simple code functions.

4. missing buy handler

when someone buys the cheapest NFT the token will not be reflected.

5. the resolver will kill the squid 

FOUR joins will make the server cry.

daiagi commented:
Hi @vikiival 

let's address the points
1, 2, 4

sure, np, will fix

>3.  too much repetitive code

mind point the finger at where?

> 5.the resolver will kill the squid

I understand
1. I hope you see why we need the resolver in the first place (basically just for the own filter)
2. I only see it possible to remove 1 JOIN, not considering at the moment namings what do you think of this:

```sql

WITH CheapestNFT AS (
    SELECT 
        ne.token_id,
        MIN(ne.price) AS "cheapestNFTPrice",
        COUNT(ne.id) as count
    FROM 
        nft_entity ne
    WHERE 
        ne.current_owner = $1
    GROUP BY
        ne.token_id
)

SELECT
    t.id as id,
    t.name as name,
    t.image as image,
    t.media as media,
    t.metadata as metadata,
    me.id as "metaId",
    me.description as "metaDescription",
    me.image as "metaImage",
    me.animation_url as "metaAnimationUrl",
    t.block_number as "blockNumber",
    t.created_at AS "createdAt",
    t.updated_at AS "updatedAt",
    c.count as count,
    c."cheapestNFTPrice",
    col.id AS "collectionId",
    col.name AS "collectionName"
FROM
    token_entity as t
        JOIN CheapestNFT as c ON t.id = c.token_id
        JOIN collection_entity as col ON t.collection_id = col.id
        JOIN metadata_entity as me ON t.meta_id = me.id
WHERE
    ($5::bigint IS NULL OR c."cheapestNFTPrice" >= $5::bigint) AND
    ($6::bigint IS NULL OR c."cheapestNFTPrice" > $6::bigint) AND
    ($7::bigint IS NULL OR c."cheapestNFTPrice" <= $7::bigint)
ORDER BY $4 LIMIT $2 OFFSET $3;
```

vikiival commented:
> mind point the finger at where?


![Screenshot 2023-10-06 at 12 31 27](https://github.com/kodadot/stick/assets/22471030/066139f7-0733-43d7-8cbe-b88c6ceb9da0)

![Screenshot 2023-10-06 at 12 31 35](https://github.com/kodadot/stick/assets/22471030/1d33ef6e-bb12-4b6d-ab66-b61eca9c03b5)

![Screenshot 2023-10-06 at 12 31 42](https://github.com/kodadot/stick/assets/22471030/b5caa6a0-84cd-4597-8249-a1f9a5f96bb6)

![Screenshot 2023-10-06 at 12 31 49](https://github.com/kodadot/stick/assets/22471030/53ffb430-b0ad-470e-a0f8-76966df19013)

daiagi commented:
all done @vikiival 

please check again

vikiival commented on src/mappings/uniques/destroy.ts:
why?

daiagi replied:
maybe you were no looking at all commits?
i didn't alter this file

vikiival commented on src/mappings/shared/token/buy.ts:
typo or something else ?

daiagi replied:
no. 
handling buy is exactly the same as handling list
if you prefer, I can call list handler instead of this name alias

vikiival commented on src/mappings/shared/token/list.ts:
```diff
@@ -0,0 +1,30 @@
+      warn(OPERATION, `nft ${nft.id} is not linked to a token`)
+      return
+    }
+    const cheapest = nftWithToken.token.cheapest
+    if (!nft.price && !cheapest?.price) {
+      return
```
`cheapest` will be always null as get with does lazy eval

daiagi replied:
thanks

vikiival commented on src/mappings/shared/token/tokenAPI.ts:
```diff
@@ -0,0 +1,75 @@
+  constructor(private store: Store) {}
+
+  async create(collection: CE, nft: NE): Promise<TE | undefined> {
+    const nftMedia = nft.image ?? nft.media
```
utils

vikiival commented on src/mappings/shared/token/tokenAPI.ts:
```diff
@@ -0,0 +1,75 @@
+    }
+    const tokenId = generateTokenId(collection.id, nftMedia)
+    debug(OPERATION, { createToken: `Create TOKEN ${tokenId} for NFT ${nft.id}` })
+    const tokenName = typeof nft.name === 'string' ? nft.name?.replace(/([#_]\d+$)/g, '').trim() : ''
```
utils

vikiival commented on src/mappings/shared/token/tokenAPI.ts:
```diff
@@ -0,0 +1,75 @@
+
+  async create(collection: CE, nft: NE): Promise<TE | undefined> {
+    const nftMedia = nft.image ?? nft.media
+    if (!nftMedia || nftMedia === '') {
```
utils

vikiival commented on src/mappings/shared/token/utils.ts:
```diff
@@ -0,0 +1,20 @@
+  return `${collectionId}-${md5(nftMedia)}`
+}
+
+export const mediaOf = (nft: NE): string | undefined => {
```
Maybe use this?

vikiival reviewed (changes requested):
Far better! 

some updates

daiagi commented:
@vikiival fixed utils issue

also, now i realize, the usage from FE will have to be via the resolver only
1. to simplify usage of orderBy price, that needs to check cheapest
2. to support deny list (count and supply need to reflect forbidden nfts)

vikiival commented:
> also, now i realize, the usage from FE will have to be via the resolver only

The scary one? 🥺

daiagi commented:
> The scary one? 🥺
indeed

owner filtering => count/supply change => resolver
denyList => same as owner filtering really
provides easier API for price ordering etc... but that is not a must have

it's not so scary. just join on collection and metadata and 2 CTE's

vikiival commented on src/server-extension/query/tokenEntities.ts:
```diff
@@ -7,17 +7,19 @@ export const tokenEntityByOwner = `WITH cheapest_nft AS (
     WHERE 
-        ne.current_owner = $1
+    ($1::text IS NULL OR ne.current_owner = $1) AND 
+    ($7::text[] IS NULL OR ne.issuer NOT IN (SELECT unnest($7)))
```
Didn't know about `unnest` field :octocat:

vikiival commented:
> indeed

have you ran some sample tests? how it behaves on the frontend?

daiagi commented:
>have you ran some sample tests? how it behaves on the frontend?

yes , I am currently integrating  with FE running the squid on localhost 
and improving the resolver as needed (for example need to add total count field that disregards limit and offset)

so far it works very well

 also, as mentioned here https://github.com/kodadot/stick/pull/110#discussion_r1349710622
 I did a benchmark, and the results are pleasing (sub 2ms for query)

daiagi commented:
work in progress:
1. spotted a bug with cheapest
2. collection floor price can be dropped
3. need another resolver for the total amount of tokenEntities within a search criteria (price, buy, own)


![image](https://github.com/kodadot/stick/assets/22791238/d66225e8-df06-408e-a5a0-502b1ceffe31)

daiagi commented:
> spotted a bug with cheapest

a. dropped it from schmea as it is problematic (but and but events of chepeast will require fetching all token's nft's to find the cheapest again)
b. moved it to the resolver instead

> collection floor price can be dropped

dropped, replaced by cheapest

> need another resolver for the total amount of tokenEntities within a search criteria (price, buy, own)

done

vikiival commented on src/mappings/uniques/transfer.ts:
Why comment out?

daiagi replied:
sorry didn't mean to

i comment them out when I want to start the indexr from NFTS pallet for testing this or that


I didn't realize I pushed it

vikiival commented on src/mappings/shared/token/mint.ts:
```diff
@@ -0,0 +1,20 @@
+import { OPERATION, generateTokenId, mediaOf } from './utils'
+import { TokenAPI } from './tokenAPI'
+
+export async function mintHandler(context: Context, collection: CE, nft: NE): Promise<TE | undefined> {
```
```suggestion
export async function createToken(context: Context, collection: CE, nft: NE): Promise<TE | undefined> {
```

daiagi replied:
mmm but it doesn't only create token

there is the possibility of adding to a preExsting token as well

```typescript

export async function mintHandler(context: Context, collection: CE, nft: NE): Promise<TE | undefined> {
  debug(OPERATION, { mintHandler: `Handle mint for NFT ${nft.id}` })

  const nftMedia = mediaOf(nft)
  if (!nftMedia) {
    return
  }

  const tokenApi = new TokenAPI(context.store)

  const existingToken = await getOptional<TE>(context.store, TE, generateTokenId(collection.id, nftMedia))
  return await (existingToken ? tokenApi.addNftToToken(nft, existingToken) : tokenApi.create(collection, nft))
}

```

vikiival commented on src/mappings/shared/token/setMetadata.ts:
```diff
@@ -0,0 +1,32 @@
+import { OPERATION, generateTokenId, mediaOf } from './utils'
+import { TokenAPI } from './tokenAPI'
+
+export async function setMetadataHandler(context: Context, collection: CE, nft: NE): Promise<TE | undefined> {
```
```suggestion
export async function updateTokenMetada(context: Context, collection: CE, nft: NE): Promise<TE | undefined> {
```

daiagi replied:
here too

it doesn't up[date the token metadata

it removed the nft from the token
and creates a new one or adds it to a previously exiting token

basically move the nft (on which the metadata changed) from 1 token to another

vikiival reviewed (changes requested):
smol

daiagi commented:
@vikiival 
left some comments

let me know if you still want me to change the handlers names

Thank you :)

vikiival commented:
> let me know if you still want me to change the handlers names

Like this is ok

vikiival reviewed (approved):
Let it roll
